cranelift/codegen/meta/src/gen_inst.rs

```diff
@@ -1179,6 +1416,12 @@ pub(crate) fn generate(
     gen_try_from(all_inst, &mut fmt);
     fmt.update_file(opcode_filename, out_dir)?;
 
+    // ISLE DSL.
+    let mut fmt = Formatter::new();
+    gen_isle(&formats, all_inst, &mut fmt);
+    let crate_src_dir = crate_dir.join("src");
+    fmt.update_file(isle_filename, &crate_src_dir.display().to_string())?;
```

Comment:
This unconditionally writes to the source dir. Crates.io dependencies aren't allowed to write to their source dir. `cargo publish` checks that the source dir stays identical. If the isle file is up to date, this check will currently pass, but if in the future the source dirs are made readonly, this will break.

Reply:
Maybe write it to `OUT_DIR` too and let ISLE read it from there?

Reply:
I wonder if the checked-in-generated-code approach could be done as follows: (i) provide a shell-script that regenerates all relevant source; (ii) in CI, fail if generated source is not up-to-date; (iii) do nothing in build.rs, to avoid surprising side-effects in the source tree.

This seems to be the simplest and fastest option, and keeps to the general principle that the source tree is human-managed and not edited by tools except by explicit invocation of a command. It does however require a manual `./update-isle.sh && cargo build` when iterating on the rules. Maybe a way to avoid a footgun here is to error in `build.rs` if the generated source is out of data. We could do *that* without putting `islec` in the build deps by recording a hash of the ISLE source alongside generated code, and checking against that hash in build.rs. Thoughts?

Reply:
I plan to add a CI check for whether the ISLE code is up-to-date, will implement that soon.

I think it would be easier to add a `rebuild-isle` feature to `cranelift-codegen-meta` than to do all of that with the hashing and everything.

Reply:
I think, at least the way I'm imagining it, the rebuild (`rebuild-isle` feature sounds good, I guess that's an implementation strategy for an `update-isle.sh` or moral equivalent?) and the local-development-loop check (using hashing or something else) are orthogonal.

Basically, (i) the rebuild is an explicit action that the user invokes in order to update the in-tree generated source, and (ii) hashing and checking in `build.rs` is one strategy of ensuring that a local build doesn't succeed unless the user does this explicit update.

My thinking comes from two principles -- first, that we shouldn't have a setup where an edit to a `.isle` file is ignored during a rebuild if one forgets to do a manual step (otherwise, we're bound to waste time debugging why a change didn't make a difference...) and second, that it's kind of an anti-pattern to have a build step that *automatically* updates a file in `src/`, without the user asking for it. So from those two principles, we have that (i) there should be a separate update step, and (ii) we should check that it was done.

Possibly those aren't principles that we want, but they seem to me to be a reasonable starting place; thoughts?

Reply:
> So from those two principles, we have that (i) there should be a separate update step, and (ii) we should check that it was done.

Agreed with these principals.

I'm not convinced we have to do more than what this PR currently does in *this* PR though.

Like I think there is a separate chunk of work that removes all of `cranelift-codegen-meta` from `cranelift-codegen`'s dependency tree and does the all things you're talking about. But for now, the ISLE integration basically does what we already do for the rest of `cranelift-codegen-meta`, and I think that is fine for landing the initial ISLE integration.

Reply:
Ah, OK, possibly I'm not understanding then -- from what I can tell in this PR, the `rebuild-isle` feature is non-default, and if I (i) do a clean git-clone, (ii) modify some ISLE source, and (iii) `cargo build`, without specifying the feature or doing anything else, my local build doesn't pick up the new ISLE changes -- is that right? Or is there some plumbing somewhere that takes care of that?

This is basically the case that motivates what I had suggested above -- I think that we need a check not just in CI, but locally, that the checked-in code matches what would have been generated (but only actually overwrite if the user opts-in, with a feature or script or whatnot). To do otherwise seems to me to invite lots of confusion the first time a new contributor tries to add a lowering and wonders why it doesn't work :-)

Reply:
Yeah if you don't build with `--features rebuild-isle` then it won't rebuild ISLE or tell you to use that feature right now.

I think this is fine for this PR (since we **do** check for and catch this in CI in this PR) and then in a future PR we can redo the whole paradigm for how we do meta, compiler compile time.

Reply:
OK; I agree it's fine for now (first PR) but it seems like something that's likely to bite us pretty soon after, just as soon as others start to use it. I suspect it could be addressed in ~50 lines or so of code in `build.rs`, without requiring us to build the ISLE compiler in the default build, just hashing files and either reading/comparing or writing to a checked-in "manifest". I'm happy to do so once the initial thing merges. In the meantime could you add a TODO somewhere with a tracking issue?

Thanks for the patience in working this one out!

Reply:
https://github.com/bytecodealliance/wasmtime/issues/3508